### PR TITLE
Msm/update docs

### PR DESCRIPTION
**Pull request by HadarIngonyama**
Updates MSM documentation

### CODE REVIEW

File: docs/docs/icicle/golang-bindings/msm.md [hunk 1]
```diff
@@ -152,6 +152,10 @@ out.Malloc(batchSize*p.Size(), p.Size())
 ...
 ```
 
+## Parameters for optimal performance
+
+Please refer to the [primitive description](../primitives/msm.md#choosing-optimal-parameters)
```
Comment (jeremyfelder) [inline docs/docs/icicle/golang-bindings/msm.md:157]: When the site is built the path drops the file extension

```suggestion
Please refer to the [primitive description](../primitives/msm#choosing-optimal-parameters)
```

File: docs/docs/icicle/primitives/msm.md [hunk 2]
```diff
@@ -54,36 +54,142 @@ You can learn more about how MSMs work from this [video](https://www.youtube.com
...
+## Algorithm description
+
+We follow the bucket method algorithm. The GPU implementation consists of four phases:
+
+1. Preparation phase - The scalars are split into smaller scalars of `c` bits each. These are the bucket indices. The points are grouped according to their corresponding bucket index and the buckets are sorted by size.
+2. Accumulation phase - Each bucket accumulates all of its points using a single thread. More than one thread is assigned to large buckets, in proportion to their size. A bucket is considered large if its size is above the large bucket threshold that is determined by the `large_bucket_factor` parameter. The large bucket threshold is the expected average bucket size times the `large_bucket_factor` parameter.
+3. Buckets Reduction phase - bucket results are multiplied by their corresponding bucket number and each bucket module is reduced to a small number of final results. By default this is done by and iterative algorithm which is highly parallel. Setting `is_big_triangle` to `true` will switch this phase to the running sum algorithm described in the above YouTube talk which is much less parallel.
```
Comment (jeremyfelder) [inline docs/docs/icicle/primitives/msm.md:63]:
```suggestion
3. Buckets Reduction phase - bucket results are multiplied by their corresponding bucket number and each bucket module is reduced to a small number of final results. By default, this is done by an iterative algorithm which is highly parallel. Setting `is_big_triangle` to `true` will switch this phase to the running sum algorithm described in the above YouTube talk which is much less parallel.
```

File: docs/docs/icicle/rust-bindings/msm.md [hunk 3]
```diff
@@ -144,6 +144,10 @@ msm::msm(&scalars, &points, &cfg, &mut msm_results).unwrap();
 
 Here is a [reference](https://github.com/ingonyama-zk/icicle/blob/77a7613aa21961030e4e12bf1c9a78a2dadb2518/wrappers/rust/icicle-core/src/msm/mod.rs#L108) to the code which automatically sets the batch size. For more MSM examples have a look [here](https://github.com/ingonyama-zk/icicle/blob/77a7613aa21961030e4e12bf1c9a78a2dadb2518/examples/rust/msm/src/main.rs#L1).
 
+## Parameters for optimal performance
+
+Please refer to the [primitive description](../primitives/msm.md#choosing-optimal-parameters)
```
Comment (jeremyfelder) [inline docs/docs/icicle/rust-bindings/msm.md:149]:
```suggestion
Please refer to the [primitive description](../primitives/msm#choosing-optimal-parameters)
```

File: docs/docs/icicle/primitives/msm.md [hunk 4]
```diff
@@ -54,36 +54,142 @@ You can learn more about how MSMs work from this [video](https://www.youtube.com
...
+The two most important parameters for performance are `c` and the `precompute_factor`. They affect the number of EC additions as well as the memory size. When the points are not known in advance we cannot use precomputation. In this case the best `c` value is usually around $log_2(msmSize) - 4$. However, in most protocols the points are known in advanced and precomputation can be used unless limited by memory. Usually it's best to use maximum precomputation (such that we end up with only a single bucket module) combined we a `c` value around $log_2(msmSize) - 1$.
+
+## Memory usage estimation
+
+The main memory requirements of the MSM are the following:
+
+Scalars - `sizeof(scalar_t) * msm_size * batch_size`
+Scalar indices - `~6 * sizeof(unsigned) * nof_bucket_modules * msm_size * batch_size`
+Points - `sizeof(affine_t) * msm_size * precomp_factor * batch_size`
+Buckets - `sizeof(projective_t) * nof_bucket_modules * 2^c * batch_size`
```
Comment (jeremyfelder) [inline docs/docs/icicle/primitives/msm.md:143]: If we want these to be rendered on separate lines then either add a <br> at the end or a new line between each

File: docs/docs/icicle/primitives/msm.md [hunk 5]
```diff
@@ -54,36 +54,142 @@ You can learn more about how MSMs work from this [video](https://www.youtube.com
...
+```
+
+Here are the parameters and the results for the different cases:
+
+| MSM size | Batch size | Precompute factor | c | Memory estimation (GB) | Actual memory (GB) | Single MSM time (ms) |
+| --- | --- | --- | --- | --- | --- | --- |
+| 10 | 1 | 1 | 9 | 0.00227 | 0.00277 | 9.2 |
+| 10 | 1 | 23 | 11 | 0.00227 | 0.00272 | 1.76 |
+| 10 | 1000 | 1 | 7 | 0.00227 | 1.09 | 0.051 |
+| 10 | 1000 | 23 | 11 | 0.00227 | 2.74 | 0.025 |
```
[the remark this answers is not in the document]
Comment (HadarIngonyama) [reply docs/docs/icicle/primitives/msm.md:183]: true, i missed that, fixing..

File: docs/docs/icicle/primitives/msm.md (hunk 4 above)
Comment (HadarIngonyama) [reply docs/docs/icicle/primitives/msm.md:143]: add what?

Comment (jeremyfelder) [reply docs/docs/icicle/primitives/msm.md:143]: 😅 not quoting the `<br>` caused a line break

File: docs/docs/icicle/primitives/msm.md [hunk 6]
```diff
@@ -54,36 +54,142 @@ You can learn more about how MSMs work from this [video](https://www.youtube.com
...
+
+$max(scalars + scalarIndices, scalars + points + buckets)$
+
+This gives a good approximation within 10% of the actual required memory for most cases.
+
+## Example parameters
+
+Here is a useful table showing optimal parameters for different MSMs. They are optimal for BLS12-377 curve when running on NVIDIA GeForce RTX 3090 Ti. This is the configuration used:
+
+```c++
```
Comment (LeonHibnik) [inline docs/docs/icicle/primitives/msm.md:157]:
```suggestion
```cpp
```

File: docs/docs/icicle/primitives/msm.md [hunk 7]
```diff
@@ -54,36 +54,142 @@ You can learn more about how MSMs work from this [video](https://www.youtube.com
...
+3. Buckets Reduction phase - bucket results are multiplied by their corresponding bucket number and each bucket module is reduced to a small number of final results. By default, this is done by an iterative algorithm which is highly parallel. Setting `is_big_triangle` to `true` will switch this phase to the running sum algorithm described in the above YouTube talk which is much less parallel.
+4. Final accumulation phase - The final results from the last phase are accumulated using the double-and-add algorithm.
+
+## Batched MSM
+
+The MSM supports batch mode - running multiple MSMs in parallel. It's always better to use the batch mode instead of running single msms in serial as long as there is enough memory available. We support running a batch of MSMs that share the same points as well as a batch of MSMs that use different points.
+
+## MSM configuration
+
+```c++
```
Comment (LeonHibnik) [inline docs/docs/icicle/primitives/msm.md:72]:
```suggestion
```cpp
```

File: docs/docs/icicle/primitives/msm.md [hunk 8]
```diff
@@ -54,36 +54,142 @@ You can learn more about how MSMs work from this [video](https://www.youtube.com
...
+    false,          // are_scalars_montgomery_form
+    true,           // are_points_on_device
+    false,          // are_points_montgomery_form
+    true,           // are_results_on_device
+    false,          // is_big_triangle
+    true            // is_async
+  };
+```
+
+Here are the parameters and the results for the different cases:
```
Comment (LeonHibnik) [inline docs/docs/icicle/primitives/msm.md:176]: Should add the test setup

**Review by LeonHibnik — COMMENTED** [review]
Looks good, left minor comments

**Review by LeonHibnik — APPROVED** [review]

**Review by jeremyfelder — APPROVED** [review]

**Pull request merged**